### PR TITLE
Move LoadAddrman from init to addrdb

### PR DESCRIPTION
**Pull request by maflcko**
Init should only concern itself with the initialization order, not the detailed initialization logic of every module.

Discussed in https://github.com/bitcoin/bitcoin/pull/22697#discussion_r691310217

### CODE REVIEW

File: src/addrman.cpp [hunk 1]
```diff
@@ -733,3 +736,22 @@ std::vector<bool> CAddrMan::DecodeAsmap(fs::path path)
     }
     return bits;
 }
+
+void LoadAddrman(std::unique_ptr<CAddrMan>& addrman, const ArgsManager& args)
+{
+    assert(!addrman);
```
Comment (laanwj) [inline src/addrman.cpp:742]: Maybe return `std::unique_ptr<CAddrMan>`? (or an optional, if it can fail, but doesn't seems so)

**Comment by vasild** [comment]
Concept ACK

File: src/addrman.cpp [hunk 2]
```diff
@@ -733,3 +736,22 @@ std::vector<bool> CAddrMan::DecodeAsmap(fs::path path)
...
+    auto check_addrman = std::clamp<int32_t>(args.GetArg("-checkaddrman", DEFAULT_ADDRMAN_CONSISTENCY_CHECKS), 0, 1000000);
+    addrman = std::make_unique<CAddrMan>(/* deterministic */ false, /* consistency_check_ratio */ check_addrman);
+
+    // Load addresses from peers.dat
+    int64_t nStart = GetTimeMillis();
```
Comment (jonatack) [inline src/addrman.cpp:747]: if you retouch
```suggestion
    const int64_t nStart{GetTimeMillis()};
```

**Review by jonatack — COMMENTED** [review]
Concept ACK

File: src/addrman.cpp (hunk 1 above)
Comment (MarcoFalke) [reply src/addrman.cpp]: Done

File: src/addrman.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/addrman.cpp]: Done

**Comment by jnewbery** [comment]
I'm ~0 on this change. I don't think there's any real benefit in moving this initialization code from init.cpp to a free function in addrdb.cpp that gets called by init. That doesn't seem to me to be reducing the complexity of what `AppInitMain()` is doing, rather than just hiding it behind a function call.

**Comment by MarcoFalke** [comment]
init.cpp is almost 2000 lines of code. I don't see why init.cpp needs to include all existing headers to learn about the implementation details of each module to initialize them. Just like wallet initialization and block importing was moved out, I think the address database details can be moved out as well.

**Comment by jnewbery** [comment]
> I don't see why init.cpp needs to include all existing headers to learn about the implementation details of each module to initialize them.

This PR doesn't reduce the number of headers that init.cpp includes (and I also don't think it's a problem for init to include lots of headers - after all, it's init's job to start everything up so it needs to include the headers for all the submodules).

> Just like wallet initialization and block importing was moved out, I think the address database details can be moved out as well.

The difference with wallet initialization and block importing is that those are internal to the wallet and validation modules. Addrman initialization from peers.dat is something that's being done to a CAddrMan externally, even after this PR.

To be clear, I don't object to this and I'm certainly not NACKing, I just don't think it brings much benefit.

**Comment by MarcoFalke** [comment]
As this is required for some other changes, I am just going to bundle with them.

**Pull request closed without merging**